### PR TITLE
Feat/hmac token signing

### PR DESCRIPTION
**Pull request by psibean**
The eventual intention before the next major release is to completely revamp the documentation and the testing suite as well, as well as provide more and better detailed examples. For testing, instead of repeating the same suite for all of the different configurations, I'd like to start splitting out the tests and only run the minimal amount needed, as well as generate a test coverage report to actually capture what has and hasn't been tested.

These changes are implicitly dependent on https://github.com/Psifi-Solutions/csrf-csrf/pull/60 

The documentation and test changes aren't the best and there are some new edge cases not covered in the tests, such as one session trying to make a request with a CSRF token generated for another session. However, I plan on overhauling both the documentation and the tests before the next release, hopefully generating some sort of documentation from TSDoc, and also including a test coverage report.

### CODE REVIEW

**Review by davidgonmar — COMMENTED** [review]
lgtm.
I am assuming if one does not want to use session identifiers they can just do getSessionIdentifier = () => "", right?
Is it the best option to use express session as the default though?

**Comment by psibean** [comment]
> lgtm. I am assuming if one does not want to use session identifiers they can just do getSessionIdentifier = () => "", right?

I think the direction I'd like to go is to not explicitly support cases where the session id is not tied to the token. The recommended practice is that generated tokens should only be valid for the session that requested it, and this acheives that in a stateless way. In the case where `getSessionIdentifier` returns "", yeah that would work, but it will also be recommended that you otherwise tie the token to the session by some other means.

> Is it the best option to use express session as the default though?

Not sure, are you suggesting we don't provide a default, thus removing the dev dependency on express-session?
And users _have_ to provide this method? I wouldn't want to put a return of "" as a default, so no default would be the next best thing and would alleviate the implicit express-session introduced here. Could update doc to making it a required config option.

**Comment by davidgonmar** [comment]
> > lgtm. I am assuming if one does not want to use session identifiers they can just do getSessionIdentifier = () => "", right?
> 
> I think the direction I'd like to go is to not explicitly support cases where the session id is not tied to the token. The recommended practice is that generated tokens should only be valid for the session that requested it, and this acheives that in a stateless way. In the case where `getSessionIdentifier` returns "", yeah that would work, but it will also be recommended that you otherwise tie the token to the session by some other means.
> 
> > Is it the best option to use express session as the default though?
> 
> Not sure, are you suggesting we don't provide a default, thus removing the dev dependency on express-session? And users _have_ to provide this method? I wouldn't want to put a return of "" as a default, so no default would be the next best thing and would alleviate the implicit express-session introduced here. Could update doc to making it a required config option.

Hmmm isn't express-session stateful actually (afaik it stores them in memory by default)? I get the appeal, but the main point of double submit pattern is for it to be completely stateless, right?

**Comment by davidgonmar** [comment]
> > > lgtm. I am assuming if one does not want to use session identifiers they can just do getSessionIdentifier = () => "", right?
> > 
> > 
> > I think the direction I'd like to go is to not explicitly support cases where the session id is not tied to the token. The recommended practice is that generated tokens should only be valid for the session that requested it, and this acheives that in a stateless way. In the case where `getSessionIdentifier` returns "", yeah that would work, but it will also be recommended that you otherwise tie the token to the session by some other means.
> > > Is it the best option to use express session as the default though?
> > 
> > 
> > Not sure, are you suggesting we don't provide a default, thus removing the dev dependency on express-session? And users _have_ to provide this method? I wouldn't want to put a return of "" as a default, so no default would be the next best thing and would alleviate the implicit express-session introduced here. Could update doc to making it a required config option.
> 
> Hmmm isn't express-session stateful actually (afaik it stores them in memory by default)? I get the appeal, but the main point of double submit pattern is for it to be completely stateless, right?

Also, I can't really think of a stateless way of maintaining a session in that sense.

**Comment by psibean** [comment]
> Also, I can't really think of a stateless way of maintaining a session in that sense.

Right, but in a case where, let's say, a user is using JWT is a httpOnly cookie, then `getSessionIdentifier` here would just return the JWT from the cookie value. Or whatever `httpOnly` cookie value they're using for their authenticationauthorization.

If they aren't using a `httpOnly` cookie, then they aren't susceptible to CSRF, regardless of stateless or not.

I understand that the double submit pattern can be used from only the client side, without the backend being involved in the token generation at all - but all of the best practices recommend a backend using a secret so that you can guarantee where the token was generated from, and additionally, it's also recommended to include some piece of information when hashing that ties the token directly to the person who requested it.

Maybe I should rename it to `getUniqueIdentifier` as it isn't necessarily a session.

**Comment by davidgonmar** [comment]
> > Also, I can't really think of a stateless way of maintaining a session in that sense.
> 
> Right, but in a case where, let's say, a user is using JWT is a httpOnly cookie, then `getSessionIdentifier` here would just return the JWT from the cookie value. Or whatever `httpOnly` cookie value they're using for their authenticationauthorization identification.
> 
> If they aren't using a `httpOnly` cookie, then they aren't susceptible to CSRF, regardless of stateless or not.
> 
> I understand that the double submit pattern can be used from only the client side, without the backend being involved in the token generation at all - but all of the best practices recommend a backend using a secret so that you can guarantee where the token was generated from, and additionally, it's also recommended to include some piece of information when hashing that ties the token directly to the person who requested it.
> 
> Maybe I should rename it to `getUniqueIdentifier` as it isn't necessarily a session, it could also be the users id. The problem with using the user id, it doesn't exist before authenticating. But for people using JWT as a cookie, they don't typically need the authentication routes protected because there's no httpOnly cookie involved at that point for them.
> 
> Edit: I think I accidentally deleted a message, no idea how.
> 
> For express-session using memory by default, yes, it does, but express-session also says NOT to use the default in production and has a list of recommended DB or cache based stores. But yeah, it's stateful. Though providing it as a default doesn't require express-session if you're overriding it. Nor does the default even require express-session, it just requires a `session` object with an `id` to exist on the request, regardless of where it came from.

I think that's fine yeah. I saw the 'session' in the name and did not think of that haha

**Review by davidgonmar — APPROVED** [review]

**Comment by psibean** [comment]
I think I finally know which direction I want to go with `csrf-csrf` and v4.

Let's refer to the [updated HMAC example on OWASP](https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md#employing-hmac-csrf-tokens).

Key changes with this PR implementation that will need to be captured in the squash merge commit message (once merging), and will also need to be covered in the v4 upgrade guide in the changelog at release time:

* `httpOnly` now defaults to false - it is professional consensus that this is generally fine. In the upgrade guide this will be covered extensively and a link will be provided to the issue discussion regarding that one particular case. However, even in that particular case, whilst the `httpOnly` option may have mitigated some of the impact, the initial vulnerability was not the responsibility of CSRF protection. Developers who wish to maintain `httpOnly: true` behaviour can still configure it that way. Allowing consumption of the token via the cookie will make the developer experience nicer overall, it is after all how Django, Laravel, Next - and various other frameworks default the implementation. Plus, given how the hmac token works, the frontend can provide the token without having to worry about stripping out a hash.
* `getSessionIdentifier` is now required, no default provided
* `generateToken` method returned renamed to `generateCsrfToken`
* `signed` option of the `cookieOptions` is no longer supported, it's redundant, the hmac is by definition a signed value
* Various TypeScript types have been tweaked to accommodate the implementation
* In a later PR `getSessionIdentifier` will be renamed to `getUniqueIdentifier` (will be documented as part of that).

File: src/tests/testsuite.ts
Comment (psibean) [inline src/tests/testsuite.ts:1]: This lot of tests is no longer required. All tests are now testing the `getSessionIdentifier` as it is now a required option.

File: src/tests/testsuite.ts [hunk 1]
```diff
@@ -114,49 +117,42 @@ export const createTestSuite: CreateTestsuite = (name, doubleCsrfOptions) => {
...
         const { mockRequest, mockResponse, decodedCookieValue } =
-          generateMocksWithTokenIntenral();
-        // modify the cookie to make the token/hash pair invalid
-        signed
-          ? (mockRequest.signedCookies[cookieName] = `s:${sign(
-              (decodedCookieValue as string).split("|")[0] + "|invalid-hash",
-              mockRequest.secret as string,
-            )}`)
-          : (mockRequest.cookies[cookieName] =
-              (decodedCookieValue as string).split("|")[0] + "|invalid-hash");
```
Comment (psibean) [inline src/tests/testsuite.ts:141]: Signed option is no longer supported, hmac is inherently signed. Technically the previous hash implementation was also inherently signed. It's redundant, nobody needs this. Thus all tests specific to `signed` option have been removed.

File: src/index.ts [hunk 2]
```diff
@@ -52,59 +54,66 @@ export function doubleCsrf({
...
-        validateTokenAndHashPair({
-          csrfToken,
-          csrfTokenHash,
-          possibleSecrets,
-          sessionIdentifier: getSessionIdentifier(req),
-        })
-      ) {
-        // If the pair is valid, reuse it
-        return { csrfToken, csrfTokenHash };
+    if (cookieName in req.cookies && !overwrite) {
```
Comment (psibean) [inline src/index.ts:91]: Technically it should have already been checking if the cookie existed first, before getting the cookie out, this is just a more sane check. If the cookie isn't present, treat it as a new request.

File: package.json [hunk 3]
```diff
@@ -52,8 +52,8 @@
     "@types/chai": "^4.3.11",
     "@types/cookie": "^0.6.0",
     "@types/cookie-parser": "^1.4.6",
-    "@types/cookie-signature": "^1.1.2",
     "@types/express": "^4.17.21",
+    "@types/express-session": "^1.18.0",
```
Comment (psibean) [inline package.json:56]: Remove this
